File: apps/trend-fetcher/src/fetchers/youtube.ts
import type { FetchedTrend, YouTubeConfig } from './types.js'

interface YouTubeItem {
  snippet: { title: string; description: string }
  id: { videoId: string }
}

interface YouTubeResponse {
  items: YouTubeItem[]
}

export async function fetchYouTube(config: YouTubeConfig): Promise<FetchedTrend[]> {
  const apiKey = process.env.YOUTUBE_API_KEY
  if (!apiKey) {
    console.warn('[trend-fetcher/youtube] YOUTUBE_API_KEY not set, skipping')
    return []
  }

  const url = new URL('https://www.googleapis.com/youtube/v3/search')
  url.searchParams.set('part', 'snippet')
  url.searchParams.set('q', config.query)
  url.searchParams.set('type', 'video')
  url.searchParams.set('order', 'viewCount')
  url.searchParams.set('maxResults', String(config.maxResults ?? 10))
  url.searchParams.set('key', apiKey)

  try {
    const res = await fetch(url.toString(), { signal: AbortSignal.timeout(15_000) })
    if (!res.ok) {
      console.error('[trend-fetcher/youtube] HTTP %d for query "%s"', res.status, config.query)
      return []
    }
    const data = (await res.json()) as YouTubeResponse
    return data.items.map((item) => ({
      title: item.snippet.title,
      url: `https://www.youtube.com/watch?v=${item.id.videoId}`,
      ...(item.snippet.description ? { description: item.snippet.description } : {}),
    }))
  } catch (err) {
    console.error('[trend-fetcher/youtube] Error for query "%s": %o', config.query, err)
    return []
  }
}

File: workers/scrapers-py/src/contento_scrapers/sources/youtube.py
from __future__ import annotations

import hashlib
import os
import uuid
from datetime import datetime, timezone

import httpx

from contento_scrapers.models import TrendDiscovered

_YT_API_BASE = 'https://www.googleapis.com/youtube/v3'


async def scrape(workspace_id: str) -> list[TrendDiscovered]:
    api_key = os.getenv('YOUTUBE_API_KEY', '')
    if not api_key:
        return []

    region = os.getenv('YOUTUBE_REGION_CODE', 'US')
    now = datetime.now(timezone.utc).isoformat()

    params = {
        'part': 'snippet',
        'chart': 'mostPopular',
        'regionCode': region,
        'maxResults': '5',
        'key': api_key,
    }

    try:
        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.get(f'{_YT_API_BASE}/videos', params=params)
            resp.raise_for_status()
            data = resp.json()
    except Exception as exc:
        print(f'[youtube] scrape error: {exc}')
        return []

    events: list[TrendDiscovered] = []
    for item in data.get('items', []):
        video_id = item.get('id', '')
        snippet = item.get('snippet', {})
        title = snippet.get('title', '')
        description = snippet.get('description', '') or None
        url = f'https://youtube.com/watch?v={video_id}'
        trend_id = f'youtube:{hashlib.md5(url.encode()).hexdigest()[:16]}'
        events.append(
            TrendDiscovered(
                eventId=str(uuid.uuid4()),
                workspaceId=workspace_id,
                timestamp=now,
                trendId=trend_id,
                title=title,
                url=url,
                source='youtube',
                description=description,
            )
        )
    return events
